utilities/src/main/java/au/csiro/pathling/utilities/ResourceHolder.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 Commonwealth Scientific and Industrial Research
+ * Organisation (CSIRO) ABN 41 687 119 230.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package au.csiro.pathling.utilities;
+
+import static org.apache.commons.io.IOUtils.closeQuietly;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A class which can be used to manage the lifecycle of a number of resources which need to be
+ * closed.
+ * 
+ * @author Piotr Szul
+ */
+@Slf4j
+public class ResourceHolder implements Closeable {
+
+  @Nonnull
+  private final List<Closeable> resourcesToClose;
+
+
+  /**
+   * Constructs a new {@link ResourceHolder} with provided resources to close.
+   *
+   * @param resourcesToClose the resources to close
+   */
+  protected ResourceHolder(@Nonnull final Closeable... resourcesToClose) {
+    this.resourcesToClose = new ArrayList<>(Arrays.asList(resourcesToClose));
+  }
+
+  /**
+   * Adds a resource to the list of resources to close.
+   *
+   * @param resource the resource to close
+   * @param <T> the type of the resource
+   * @return the resource
+   */
+  protected <T extends Closeable> T registerResource(@Nonnull final T resource) {
+    synchronized (resourcesToClose) {
+      resourcesToClose.add(resource);
+      return resource;
+    }
+  }
+
+  @Override
+  public void close() {
+    log.debug("Closing {} resources for: {}", resourcesToClose.size(), this);
+    synchronized (resourcesToClose) {
+      for (Closeable closeable : resourcesToClose) {
+        log.debug("Closing resource: {} in: {}", closeable, this);
```

Comment:
Should I be seeing this logging if I use a Python script with something like the following?

```python
spark = (
    SparkSession.builder
    .config("spark.jars", find_jar())
    .getOrCreate()
)

pc = PathlingContext.create(
        spark,
        terminology_server_url="https://r4.ontoserver.csiro.au/fhir",
        cache_override_expiry=2_628_000,
        cache_storage_type="disk",
        cache_storage_path=".local/tx-cache"
)

pc.spark.sparkContext.setLogLevel("DEBUG")
```

Reply:
Yes, at the end:

```
23/06/20 14:34:43 INFO ShutdownHookManager: Shutdown hook called
23/06/20 14:34:43 INFO ShutdownHookManager: Deleting directory /private/var/folders/x5/cw9kr5ld00g595qv0gg7lxrw00cjzw/T/spark-17e8a874-d7f2-44cb-bb94-d8f8b83df7c9
23/06/20 14:34:43 INFO ShutdownHookManager: Deleting directory /private/var/folders/x5/cw9kr5ld00g595qv0gg7lxrw00cjzw/T/spark-5c43a9fb-16b2-4554-b414-a79937f7c20e
23/06/20 14:34:43 INFO ShutdownHookManager: Deleting directory /private/var/folders/x5/cw9kr5ld00g595qv0gg7lxrw00cjzw/T/spark-17e8a874-d7f2-44cb-bb94-d8f8b83df7c9/pyspark-a21b3294-186f-45e2-af3f-02f5c0291829
23/06/20 14:34:43 DEBUG ResourceHolder: Closing 2 resources for: au.csiro.pathling.terminology.caching.InMemoryCachingTerminologyService@6aeae867
23/06/20 14:34:43 DEBUG ResourceHolder: Closing resource: org.apache.http.impl.client.InternalHttpClient@6487e431 in: au.csiro.pathling.terminology.caching.InMemoryCachingTerminologyService@6aeae867
23/06/20 14:34:43 DEBUG PoolingHttpClientConnectionManager: Connection manager is shutting down
23/06/20 14:34:43 DEBUG DefaultManagedHttpClientConnection: http-outgoing-0: Close connection
23/06/20 14:34:43 DEBUG PoolingHttpClientConnectionManager: Connection manager shut down
23/06/20 14:34:43 DEBUG ResourceHolder: Closing resource: DefaultCacheManager DefaultCacheManager in: au.csiro.pathling.terminology.caching.InMemoryCachingTerminologyService@6aeae867
23/06/20 14:34:43 DEBUG DefaultCacheManager: Stopping cache manager DefaultCacheManager
23/06/20 14:34:43 DEBUG CacheImpl: Stopping cache lookup on DefaultCacheManager
23/06/20 14:34:43 DEBUG CacheImpl: Stopping cache org.infinispan.CONFIG on DefaultCacheManager
23/06/20 14:34:43 DEBUG CacheImpl: Stopping cache validate-code on DefaultCacheManager
23/06/20 14:34:43 DEBUG CacheImpl: Stopping cache translate on DefaultCacheManager
23/06/20 14:34:43 DEBUG CacheImpl: Stopping cache subsumes on DefaultCacheManager
23/06/20 14:34:43 DEBUG DefaultCacheManager: Stopped cache manager DefaultCacheManager
23/06/20 14:34:43 DEBUG FileSystem: FileSystem.close() by method: org.apache.hadoop.fs.FilterFileSystem.close(FilterFileSystem.java:529)); Key: (szu004 (auth:SIMPLE))@file://; URI: file:///; Object Identity Hash: 597778cb
23/06/20 14:34:43 DEBUG FileSystem: FileSystem.close() by method: org.apache.hadoop.fs.RawLocalFileSystem.close(RawLocalFileSystem.java:759)); Key: null; URI: file:///; Object Identity Hash: 77576ebb
23/06/20 14:34:43 DEBUG ShutdownHookManager: Completed shutdown in 0.122 seconds; Timeouts: 0
23/06/20 14:34:43 DEBUG ShutdownHookManager: ShutdownHookManager completed shutdown.
```